Make docker_stack adhere to standard ansible return values

parmsib opened the pull request:
The names of the various fields returned from ansible modules are e.g defined here https://docs.ansible.com/ansible/latest/reference_appendices/common_return_values.html#msg.

Adhering to this improves usability and makes use of functionality for e.g stdout_lines etc.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Added missing instances of return values, and changed name of return values
where they did not adhere to ansible standard.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
docker_stack

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
One example of why this is an improvement is that without access to std_err from the module,
there is no use for `--resolve-conf`, which only makes `docker stack deploy` print
a warning to `stderr`.
<!--- Paste verbatim command output below, e.g. before and after your change -->
Before the changes, this is an example of return values I received:
```
ok: [host1] => {
    "docker_stack": {
        "changed": false,
        "failed": false
    }
}
```
after the changes, this is another example, with a bunch of irrelevant/sensitive information removed:
```
ok: [host1] => {
    "docker_stack": {
        "changed": false,
        "failed": false,
        "rc": 0,
        "stderr": "image some-registry/some-image:3.0.1 could not be accessed on a registry to r....\n\n",
        "stderr_lines": [
            "image some-registry/some-image:3.0.1 could not be accessed on a registry to record",
            "its digest..." 
            ...
        ],
        "stdout": "Updating service service1(id: id1)\nUpdating service service2(id: id2)\n",
        "stdout_lines": [
            "Updating service service1(id: id1)",
            "Updating service service2(id: id2)"
        ]
    }
}
```

parmsib committed:
Make docker_stack adhere to standard return values

The names of the various fields returned from ansible modules are e.g defined here https://docs.ansible.com/ansible/latest/reference_appendices/common_return_values.html#msg.

Adhering to this improves usability and makes use of functionality for e.g stdout_lines etc.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -261,8 +261,8 @@ def main():
             module.fail_json(msg="docker stack up deploy command failed",
-                             out=out,
-                             rc=rc, err=err)
+                             stdout=out,
```
While the other changes are ok, this one is problematic. If a playbook or role depends on the currently named return values, it will stop working. It's better to return all four (`out`, `stdout`, `err`, `stderr`) for now, and announce in the docs and porting guide that `out` and `err` are deprecated and will be removed in Ansible 2.14 (standard deprecation cycle of four versions).

parmsib replied:
Full enclosure: I know very little about what is done with the values returned from fail_json.
As you can see in my example result from before the changes, only `changed` and `failed` were exposed to me as an ansible user. This is what prompted my changes to begin with.
My point is that I'm not sure anyone could rely on the existing return values, if they were not exposed.

dariko replied:
@parmsib 
Your example is for a succesful module run: `out`, `rc` and `stderr` were included only on module failure (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_stack.py#L263-L265 and https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_stack.py#L289-L292)

parmsib replied:
Ah, thanks for the info.
One more question: Why is specifically the changes to line 264 bad, and e.g not [283](https://github.com/ansible/ansible/blob/48af1b90ea3558b1cbe7512d587481f945b50025/lib/ansible/modules/cloud/docker/docker_stack.py#L283)? Is it because the latter is exiting, and therefore we don't expect a user to be reading the values programmatically?

felixfontein replied:
Because it renames an existing return value, instead of adding a new one. If some existing role or playbook depends on the return value being named `out` resp. `err` in case of this error, it will break. Ansible tries to avoid such breakage in minor versions.

parmsib replied:
> While the other changes are ok, this one is problematic. If a playbook or role depends on the currently named return values, it will stop working. It's better to return all four (`out`, `stdout`, `err`, `stderr`) for now, and announce in the docs and porting guide that `out` and `err` are deprecated and will be removed in Ansible 2.14 (standard deprecation cycle of four versions).

I may need some guidance on what exactly to do here. Which porting guide page do I add this to, and is there somewhere I should add a "reminder" about deprecating the old return values?

felixfontein replied:
Mention this in the changelog (`deprecated_features` section) and in the [porting guide](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/porting_guides/porting_guide_2.10.rst#deprecation-notices). And also in the module's documentation.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -274,10 +274,17 @@ def main():
         if not before_after_differences:
-            module.exit_json(changed=False)
+            module.exit_json(
+              changed=False
```
```suggestion
              changed=False,
```

felixfontein commented:
Besides the errors, you need to add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to) with `minor_changes`. I'm not sure whether this will be accepted as a backport, though.

parmsib committed:
Update lib/ansible/modules/cloud/docker/docker_stack.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
Fix under-indentation of continuation line (pep8)

Issue exposed in test here https://app.shippable.com/github/ansible/ansible/runs/146667/1/console

parmsib committed:
Don't break old playbooks/roles

Made sure the changes are only adding new variables, not removing anything existing yet.

parmsib committed:
Added comment

parmsib committed:
Minor fixes

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -287,11 +295,15 @@ def main():
-                module.exit_json(changed=True, msg=out, err=err, rc=rc)
+                module.exit_json(changed=True,
+                                 msg=out, rc=rc,
+                                 err=stderr,  # Deprecated
```
```suggestion
                                 err=err,  # Deprecated
```

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -287,11 +295,15 @@ def main():
                                  rc=rc,
-                                 err=err)
+                                 out=out, err=err,  # Deprecated
+                                 stdout=out, stderr=stderr)
```
```suggestion
                                 stdout=out, stderr=std)
```

parmsib committed:
Update lib/ansible/modules/cloud/docker/docker_stack.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
Update lib/ansible/modules/cloud/docker/docker_stack.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
minor change to docker_stack.py

parmsib committed:
Add changelog fragment for PR 63467

felixfontein commented on changelogs/fragments/63467-docker-stack-return-fix.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - docker_stack - Added stdout, stderr, and rc to return values.
```
You can use [ReST](https://en.wikipedia.org/wiki/ReStructuredText) formatting here:
```suggestion
  - docker_stack - Added ``stdout``, ``stderr``, and ``rc`` to return values.
```

parmsib committed:
Format changelog fragment

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
Add fragment about docker_stack return val depr

parmsib committed:
Add docker_stack doc note about deprecated vals

parmsib committed:
Remove whitespace in empty line

parmsib committed:
Add docker_stack depr notice to porting guide

felixfontein commented on changelogs/fragments/63467-docker-stack-return-fix.yml:
```diff
@@ -0,0 +1,5 @@
+  - docker_stack - Added ``stdout``, ``stderr``, and ``rc`` to return values.
+
+deprecated_features:
+  - docker_stack - Return values ``out`` and ``err`` have been deprecated and will be removed. Use ``stdout`` and ``stderr`` instead.
```
```suggestion
  - docker_stack - Return values ``out`` and ``err`` have been deprecated and will be removed in Ansible 2.14. Use ``stdout`` and ``stderr`` instead.
```

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.10.rst:
```diff
@@ -56,13 +56,12 @@ The following functionality will be removed in Ansible 2.14. Please update updat
 * :ref:`s3_lifecycle <s3_lifecycle_module>`: the ``requester_pays`` option will be removed. It has always been ignored by the module.
 * :ref:`s3_sync <s3_sync_module>`: the ``retries`` option will be removed. It has always been ignored by the module.
-
+* :ref:`docker_stack <docker_stack_module>` has had return values `err` and `out` deprecated. Use new `stdout` and `stderr` instead.
```
```suggestion
* The return values ``err`` and ``out`` of :ref:`docker_stack <docker_stack_module>` have been deprecated. Use ``stdout`` and ``stderr`` from now on instead.
```

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.10.rst:
```diff
@@ -56,13 +56,12 @@ The following functionality will be removed in Ansible 2.14. Please update updat
 
 * The :ref:`docker_container <docker_container_module>` module has a new option, ``container_default_behavior``, whose default value will change from ``compatibility`` to ``no_defaults``. Set to an explicit value to avoid deprecation warnings.
 
-
```
Please keep this newline :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -81,6 +81,9 @@
   - pyyaml
+
+notes:
+  - Return values C(out) and C(err) have been deprecated and will be removed. Use C(stdout) and C(stderr) instead.
```
```suggestion
  - Return values I(out) and I(err) have been deprecated and will be removed in Ansible 2.14. Use I(stdout) and I(stderr) instead.
```
In general: `I(...)` is used for option names, `C(...)` for option values

felixfontein reviewed (commented):
Looks good. Just a few more nits :)

parmsib committed:
Update changelogs/fragments/63467-docker-stack-return-fix.yml

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
Update docs/docsite/rst/porting_guides/porting_guide_2.10.rst

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
Update lib/ansible/modules/cloud/docker/docker_stack.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

parmsib committed:
Added back a missing new line

felixfontein reviewed (approved):
LGTM

felixfontein commented:
Thanks a lot for fixing this!

parmsib commented:
Thank you! Sorry for the messy pull request

felixfontein commented:
@parmsib no problem, thanks a lot for sticking with me and bringing it into shape :)
